x/feegrant/filtered_fee_test.go

```diff
@@ -0,0 +1,193 @@
+				SpendLimit: smallAtom,
+			},
+			msgs:   []string{sdk.MsgTypeURL(&call)},
+			fee:    eth,
+			accept: false,
+		},
+		"non-expired": {
+			allowance: &feegrant.BasicAllowance{
+				SpendLimit: atom,
+				Expiration: &oneHour,
+			},
+			msgs:      []string{sdk.MsgTypeURL(&call)},
+			valid:     true,
+			fee:       smallAtom,
+			blockTime: now,
+			accept:    true,
+			remove:    false,
+			remains:   leftAtom,
+		},
+		"expired": {
+			allowance: &feegrant.BasicAllowance{
+				SpendLimit: atom,
+				Expiration: &now,
+			},
+			msgs:      []string{sdk.MsgTypeURL(&call)},
+			valid:     true,
+			fee:       smallAtom,
+			blockTime: oneHour,
+			accept:    false,
+			remove:    true,
+		},
+		"fee more than allowed": {
+			allowance: &feegrant.BasicAllowance{
+				SpendLimit: atom,
+				Expiration: &oneHour,
+			},
+			msgs:      []string{sdk.MsgTypeURL(&call)},
+			valid:     true,
+			fee:       bigAtom,
+			blockTime: now,
+			accept:    false,
+		},
+		"with out spend limit": {
+			allowance: &feegrant.BasicAllowance{
+				Expiration: &oneHour,
+			},
+			msgs:      []string{sdk.MsgTypeURL(&call)},
+			valid:     true,
+			fee:       bigAtom,
+			blockTime: now,
+			accept:    true,
+		},
+		"expired no spend limit": {
+			allowance: &feegrant.BasicAllowance{
+				Expiration: &now,
+			},
+			msgs:      []string{sdk.MsgTypeURL(&call)},
+			valid:     true,
+			fee:       bigAtom,
+			blockTime: oneHour,
+			accept:    false,
+		},
+	}
+
+	for name, stc := range cases {
+		tc := stc // to make scopelint happy
+		t.Run(name, func(t *testing.T) {
+			err := tc.allowance.ValidateBasic()
+			require.NoError(t, err)
+
+			ctx := app.BaseApp.NewContext(false, ocproto.Header{}).WithBlockTime(tc.blockTime)
+
+			// create grant
+			var granter, grantee sdk.AccAddress
+			allowance, err := feegrant.NewAllowedMsgAllowance(tc.allowance, tc.msgs)
+			require.NoError(t, err)
+			grant, err := feegrant.NewGrant(granter, grantee, allowance)
+			require.NoError(t, err)
+
+			// now try to deduct
+			removed, err := allowance.Accept(ctx, tc.fee, []sdk.Msg{&call})
+			if !tc.accept {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+
+			require.Equal(t, tc.remove, removed)
+			if !removed {
+				// save the new grant
+				newGrant, err := feegrant.NewGrant(
+					sdk.AccAddress(grant.Granter),
+					sdk.AccAddress(grant.Grantee),
+					allowance)
+				require.NoError(t, err)
+
+				cdc := simapp.MakeTestEncodingConfig().Codec
+				bz, err := cdc.Marshal(&newGrant)
+				require.NoError(t, err)
+
+				// load the grant
+				var loadedGrant feegrant.Grant
+				err = cdc.Unmarshal(bz, &loadedGrant)
+				require.NoError(t, err)
+
+				newAllowance, err := newGrant.GetGrant()
+				require.NoError(t, err)
+				feeAllowance, err := newAllowance.(*feegrant.AllowedMsgAllowance).GetAllowance()
+				require.NoError(t, err)
+				assert.Equal(t, tc.remains, feeAllowance.(*feegrant.BasicAllowance).SpendLimit)
+			}
```

Comment:
I actually don't understand this block so well, still.

Can we simply do:
```go
			require.Equal(t, tc.remove, removed)
			if !removed {
				assert.Equal(t, tc.remains, allowance.Allowance.GetCachedValue().(*feegrant.BasicAllowance).SpendLimit)
			}
```

Reply:
This block is the whole point of this fix. With the simplified version, we cannot check whether the allowance has been updated correctly, because the cached value was correct even before the fix.
This patch is forcing the cached value applied to the state. We must marshal & unmarshal it, which simulates the actual process.
That was the last redundant `func()` meant for. If you don't mind, I can wrap it with `func()` again, providing meaningful name and comments.

Reply:
Sorry, there was a mistake during the modification. I think this has confused you.
`newGrant` at line 185 must be `loadedGrant`.
I will also apply the fix after your response.

Reply:
OK Understood. Yes `loadedGrant` on L185 sounds already clearer. 

Could you add some comments, e.g. around marshaling to simulate the actual process of putting in block?

> If you don't mind, I can wrap it with func() again (or a block with curly brackets suffice?), providing meaningful name and comments.

I don't think that's idiomatic go. Using newlines and comments should be enough to describe what you're achieving.

Reply:
```
if !removed {
	assert.Equal(t, tc.remains, allowance.GetAllowance().(*feegrant.BasicAllowance).SpendLimit)
}
```
I align with @amaurym here, this thing works I guess, can you verify that this assert should be fail with the earlier code and passes with the changes in this PR. if that works no need of creating the Grant and marshals.

Reply:
I verified that the simplified version does not fail with both of the earlier code and the new code. And I also verified that the longer version with marshaling fails with the earlier code.